src/Host/Monitor/Impl/CredentialManager.cs

```diff
@@ -0,0 +1,388 @@
+                }
+
+                return credStatus;
+            });
+        }
+
+        private const int CRED_MAX_USERNAME_LENGTH = 256;
+        private const int CRED_MAX_DOMAIN_LENGTH = 256;
+        private const int CRED_MAX_PASSWORD_LENGTH = 256;
+        public static Task GetAndSaveCredentialsFromUserAsync() {
+            return Task.Run(() => {
+                CredUIInfo credInfo = new CredUIInfo();
+                credInfo.cbSize = Marshal.SizeOf(credInfo);
+                credInfo.pszCaptionText = Resources.Text_RHostBrokerCredentials;
+                credInfo.pszMessageText = Resources.Text_RHostBrokerCredentialsDetail;
+                credInfo.hwndParent = Process.GetCurrentProcess().MainWindowHandle;
+
+                int errorCode = 0;
+                uint authenticationPackage = 0;
+                IntPtr credentialBuffer;
+                uint credentialSize;
+                bool saveCreds = false;
+
+                int attempts = 0;
+                while (attempts < 3) {
+                    ++attempts;
+                    uint ret = CredUIPromptForWindowsCredentials(
+                        ref credInfo,
+                        errorCode,
+                        ref authenticationPackage,
+                        IntPtr.Zero,
+                        0,
+                        out credentialBuffer,
+                        out credentialSize,
+                        ref saveCreds,
+                        PromptForWindowsCredentialsFlags.CREDUIWIN_GENERIC);
+
+                    // ERROR_CANCELLED: User clicked the cancel button.
+                    if (ret == 1223) {
+                        break;
+                    }
+
+                    // ERROR_SUCCESS: User entered credentials
+                    if (ret == 0) {
+                        StringBuilder username = new StringBuilder(CRED_MAX_USERNAME_LENGTH);
+                        StringBuilder domain = new StringBuilder(CRED_MAX_USERNAME_LENGTH);
+                        StringBuilder password = new StringBuilder(CRED_MAX_USERNAME_LENGTH);
+
+                        int usernameLen = username.Capacity;
+                        int domainLen = domain.Capacity;
+                        int passwordLen = password.Capacity;
+
+                        if (!CredUnPackAuthenticationBuffer(
+                            0,
+                            credentialBuffer,
+                            credentialSize,
+                            username,
+                            ref usernameLen,
+                            domain,
+                            ref domainLen,
+                            password,
+                            ref passwordLen)) {
+                            throw new Win32Exception(Marshal.GetLastWin32Error());
+                        }
+
+                        IntPtr token = IntPtr.Zero;
+                        try {
+                            var usernameBldr = new StringBuilder(257);
```

Comment:
(or maybe it was a different constant... but either way, there should be a named constant).